### PR TITLE
feat: cross-platform force-kill primitive for stuck PHP threads

### PR DESCRIPTION
**Pull request by nicolas-grekas**
First step of the split suggested in #2287: land the force-kill
infrastructure as a standalone, reviewable primitive independent of
background workers.

## Design

Each PHP thread, at boot from its own TSRM context, hands a
`force_kill_slot` (pointers to its `EG(vm_interrupt)` and `EG(timed_out)`
atomic bools, plus `pthread_t` / Windows `HANDLE`) back to Go via
`go_frankenphp_store_force_kill_slot`. The slot lives on `phpThread`
and is protected by a per-thread `RWMutex` so the zero-and-release path
at thread exit cannot race an in-flight kill. From any goroutine, Go
passes the slot back to `frankenphp_force_kill_thread`, which stores
`true` into both atomic bools (waking the VM at the next opcode
boundary, routing through `zend_timeout` -> "Maximum execution time
exceeded") and delivers a platform-specific wake-up:

- **Linux/FreeBSD**: `pthread_kill(SIGRTMIN+3)` with a no-op handler
  installed once via `pthread_once`, `SA_ONSTACK`, no `SA_RESTART`.
  Signal delivery returns any in-flight blocking syscall with `EINTR`.
- **Windows**: `CancelSynchronousIo` + `QueueUserAPC` covers alertable
  I/O and `SleepEx`. Non-alertable `Sleep` (including PHP's `usleep`)
  stays uninterruptible.
- **macOS**: atomic-bool path only; threads stuck in blocking syscalls
  wait for the syscall to complete naturally.

**Reserved signal**: `SIGRTMIN+3`. A PHP script that calls
`pcntl_signal(SIGRTMIN+3, ...)` clobbers this. Embedders whose own Go
code uses `SIGRTMIN+3` must patch it here. glibc NPTL reserves
`SIGRTMIN..SIGRTMIN+2`, so the offset cannot go lower.

## Drain integration

`drainWorkerThreads` waits `drainGracePeriod` (30s) for each thread to
reach `Yielding`, then arms force-kill on stragglers and **keeps
waiting** until they yield. `phpThread.shutdown` does the same. There
is no abandon path: if a thread is stuck in a syscall force-kill cannot
interrupt (macOS, Windows non-alertable Sleep), the drain blocks until
the syscall returns naturally — matching pre-patch behaviour exactly,
just typically much faster because force-kill cuts a `sleep(60)` down
to milliseconds. Operators that want a harder bound rely on their
orchestrator (systemd, k8s, supervisord) to SIGKILL the process.

`go_frankenphp_on_thread_shutdown` runs on both the healthy path and
the unhealthy-during-Shutdown path so `state.Done` is set even when
force-kill bails the thread. Without it, `phpThread.shutdown`'s
`WaitFor(state.Done)` would never unblock.

## Testing

`TestRestartWorkersForceKillsStuckThread` drives the full path via a
marker file so `RestartWorkers` only arms once the worker is proven
parked in `sleep()`, then asserts bounded elapsed time and that the
post-sleep echo never runs.

### CODE REVIEW

File: frankenphp.c [hunk 1]
```diff
@@ -92,6 +92,108 @@ static bool is_forked_child = false;
...
+    its.it_value.tv_sec = 0;
+    its.it_value.tv_nsec = 1;
+    its.it_interval.tv_sec = 0;
+    its.it_interval.tv_nsec = 0;
+    timer_settime(thread_php_timers[idx], 0, &its, NULL);
+  }
+#elif defined(PHP_WIN32)
+  if (thread_handles && thread_handle_saved[idx]) {
+    CancelSynchronousIo(thread_handles[idx]);
+    QueueUserAPC((PAPCFUNC)frankenphp_noop_apc, thread_handles[idx], 0);
```
Comment (henderkes) [inline frankenphp.c]: TIL php sleep uses SleepEx (which is alertable), but usleep does not. This isn't the greatest option to wake up sleeping threads.

File: frankenphp.c [hunk 2]
```diff
@@ -92,6 +92,108 @@ static bool is_forked_child = false;
 static void frankenphp_fork_child(void) { is_forked_child = true; }
 #endif
 
+/* Best-effort force-kill for PHP threads stuck in blocking calls after the
+ * graceful-drain grace period. Cross-platform primitives with no dependency
+ * on any specific worker type; the Go side decides when to arm them.
+ * - Linux/FreeBSD ZTS: arm PHP's per-thread timer -> "max execution time" fatal
+ * - Windows: CancelSynchronousIo + QueueUserAPC -> interrupts I/O and sleeps
+ * - macOS/other: no-op (threads abandoned, exit when the blocking call returns)
```
Comment (henderkes) [inline frankenphp.c]: that's very unfortunate, but I don't know anything about macOS, so I'm not sure if it's realistic to implement anything, especially when we're leaning on non-existent zend execution timers

File: frankenphp.c [hunk 3]
```diff
@@ -92,6 +92,108 @@ static bool is_forked_child = false;
...
+
+void frankenphp_init_force_kill(int num_threads) {
+  force_kill_num_threads = num_threads;
+#ifdef ZEND_MAX_EXECUTION_TIMERS
+  thread_php_timers = calloc(num_threads, sizeof(timer_t));
+  thread_php_timer_saved = calloc(num_threads, sizeof(bool));
+#elif defined(PHP_WIN32)
+  thread_handles = calloc(num_threads, sizeof(HANDLE));
+  thread_handle_saved = calloc(num_threads, sizeof(bool));
+#endif
```
Comment (henderkes) [inline frankenphp.c]: if calloc fails we run into a segfault later

File: frankenphp.c [hunk 4]
```diff
@@ -92,6 +92,108 @@ static bool is_forked_child = false;
...
+  if (thread_php_timers && EG(pid)) {
+    thread_php_timers[idx] = EG(max_execution_timer_timer);
+    thread_php_timer_saved[idx] = true;
+  }
+#elif defined(PHP_WIN32)
+  if (thread_handles) {
+    DuplicateHandle(GetCurrentProcess(), GetCurrentThread(),
+                    GetCurrentProcess(), &thread_handles[idx], 0, FALSE,
+                    DUPLICATE_SAME_ACCESS);
+    thread_handle_saved[idx] = true;
```
Comment (henderkes) [inline frankenphp.c]: missing safety check if it actually succeeded

File: worker.go [hunk 5]
```diff
@@ -201,7 +209,31 @@ func drainWorkerThreads() []*phpThread {
...
+	case <-time.After(drainGracePeriod):
+		for _, thread := range drainedThreads {
+			if !thread.state.Is(state.Yielding) {
+				C.frankenphp_force_kill_thread(C.uintptr_t(thread.threadIndex))
+			}
+		}
+		if globalLogger.Enabled(globalCtx, slog.LevelWarn) {
+			globalLogger.LogAttrs(globalCtx, slog.LevelWarn, "worker threads did not yield within grace period, force-killing stuck threads")
+		}
+		<-done
```
Comment (henderkes) [inline worker.go:226]: circling back to Windows only actually interrupting SleepEx, wouldn't we block indefinitely here if the force_kill_thread didn't work?

File: frankenphp.c [hunk 6]
```diff
@@ -92,6 +92,108 @@ static bool is_forked_child = false;
...
+  }
+#ifdef ZEND_MAX_EXECUTION_TIMERS
+  if (thread_php_timers && thread_php_timer_saved[idx]) {
+    struct itimerspec its;
+    its.it_value.tv_sec = 0;
+    its.it_value.tv_nsec = 1;
+    its.it_interval.tv_sec = 0;
+    its.it_interval.tv_nsec = 0;
+    timer_settime(thread_php_timers[idx], 0, &its, NULL);
+  }
```
Comment (AlliBalliBaba) [inline frankenphp.c]: Instead of storing timers like this, I'd actually prefer if we just stored a reference to the execution globals of each thread at startup. You then can just set `EG(timed_out)` to true.

Comment (nicolas-grekas) [reply frankenphp.c]: Done. Each thread now saves pointers to its own `EG(vm_interrupt)` and `EG(timed_out)` at boot; `force_kill_thread` does atomic-stores into both. That path is the cross-platform core, including macOS and Windows.

We kept a platform-specific wake-up on top: `pthread_kill(SIGRTMIN+3)` on Linux/FreeBSD, `CancelSynchronousIo` + `QueueUserAPC` on Windows. Without it, a thread stuck in a blocking syscall (select, sleep, DB wait) would only be interrupted when the syscall returned on its own.

Side effect of the switch: the old timer path silently did nothing for bg workers because `EG(pid)` was zero when `max_execution_time=0`, so this also fixes a latent bug.

File: frankenphp.c (hunk 2 above)
Comment (nicolas-grekas) [reply frankenphp.c]: No longer depends on `ZEND_MAX_EXECUTION_TIMERS`. `EG(vm_interrupt)` exists on every ZTS thread, so macOS now gets VM-loop interruption too. Syscall interruption still doesn't work there (no realtime signals exposed to user-space), but that wasn't working before either.

File: frankenphp.c (hunk 3 above)
Comment (nicolas-grekas) [reply frankenphp.c]: Fixed: `frankenphp_init_force_kill` now returns with `force_kill_num_threads = 0` on allocation failure; subsequent register/kill calls become no-ops. Graceful drain is unaffected.

File: frankenphp.c (hunk 4 above)
Comment (nicolas-grekas) [reply frankenphp.c]: Added: on Windows, a failed `DuplicateHandle` now leaves the slot's `HANDLE` as `NULL` and `force_kill_thread` skips the Win32 primitives for that thread (only the atomic bools fire). No uninitialized dereference.

File: frankenphp.c (hunk 1 above)
Comment (nicolas-grekas) [reply frankenphp.c]: The concern holds on Windows for non-alertable `Sleep()`, which is what PHP's `usleep` uses. From our layer there's no fix: `TerminateThread` would leave locks in undefined state. `<-done` in `drainWorkerThreads` will wait until the syscall returns naturally in that case. Documented the limitation next to the primitive.

On Linux/FreeBSD the new `pthread_kill(SIGRTMIN+3)` path does interrupt `usleep` (and the rest) via EINTR. On macOS it's the same limitation as Windows `usleep` (no realtime signals).

One thought: this could be fixed upstream by having php-src switch `usleep` on Windows from `Sleep` to `SleepEx(msec, TRUE)`. That would make it alertable, and our `QueueUserAPC` would then interrupt it naturally. Might be worth a small PR to php-src.

File: worker.go (hunk 5 above)
Comment (nicolas-grekas) [reply worker.go:226]: Valid concern. On Windows with PHP `usleep` (non-alertable `Sleep`), `<-done` will wait until the sleep returns on its own because the APC cannot interrupt it. Documented. Fixing it properly would need php-src to use `SleepEx(msec, TRUE)` for `usleep` so the APC can wake it.

File: frankenphp.c (hunk 1 above)
Comment (henderkes) [reply frankenphp.c]: https://github.com/php/php-src/pull/21846

Comment (nicolas-grekas) [reply frankenphp.c]: Thanks! Once that lands, our `QueueUserAPC` path on Windows will interrupt `usleep` naturally with no change on our side, and the drain-indefinitely concern on `worker.go` goes away for the `usleep` case too.

File: worker.go (hunk 5 above)
Comment (henderkes) [reply worker.go:226]: Well we could technically force kill a thread. But I doubt that'd be safe.

File: frankenphp.c (hunk 1 above)
Comment (nicolas-grekas) [reply frankenphp.c]: (LLM posted this on its own but I'm sincerely grateful for this : D)

File: worker.go (hunk 5 above)
Comment (nicolas-grekas) [reply worker.go:226]: Yeah, `TerminateThread` can abandon held locks mid-modification and corrupt the process - better to lean on your php-src fix.

File: frankenphp.c [hunk 7]
```diff
@@ -92,6 +92,143 @@ static bool is_forked_child = false;
...
+static force_kill_slot *thread_slots = NULL;
+
+void frankenphp_init_force_kill(int num_threads) {
+  thread_slots = calloc((size_t)num_threads, sizeof(force_kill_slot));
+  if (thread_slots == NULL) {
+    /* Out of memory at startup: leave force-kill disabled rather than crash
+     * later in register/kill. Graceful drain still works via the yielding
+     * path. */
+    force_kill_num_threads = 0;
+    return;
```
Comment (AlliBalliBaba) [inline frankenphp.c]: I think it would be cleaner to just store the force_kill_slot on the go side in the `phpThread` struct and then pass it back to C in `frankenphp_force_kill_thread`.

Comment (nicolas-grekas) [reply frankenphp.c]: Done. The slot now lives on `phpThread` and gets populated by `go_frankenphp_store_force_kill_slot` from the PHP thread's own TSRM context at boot; `force_kill_thread(slot)` takes it by value. That lets us drop the C-side `thread_slots` array, the bounds checks, and both `init_force_kill` / `destroy_force_kill`. Signal-handler install moved to a one-shot `pthread_once` inside register; Windows `CloseHandle` lives in a new `release_thread_for_kill` called during drain. Diff is -25 lines, and happens-before is covered because the thread writes the slot before it transitions to `Ready`.

File: worker_test.go [hunk 8]
```diff
@@ -45,6 +49,56 @@ func TestWorker(t *testing.T) {
...
+	// RestartWorkers must complete within the grace period + a bit of slack.
+	// Without force-kill, it would wait for the 60s sleep to return.
+	start := time.Now()
+	frankenphp.RestartWorkers()
+	elapsed := time.Since(start)
+
+	// Grace period is 5s; allow margin for SIGALRM dispatch, PHP VM tick,
+	// and the drain's final ready.Wait() plus the restart loop.
+	const budget = 8 * time.Second
+	assert.Less(t, elapsed, budget, "drain must force-kill the stuck thread within the grace period")
```
Comment (withinboredom) [inline worker_test.go]: We should probably assert that the output from the script doesn't contain "should not reach"? That way any future changes that break the kill, but not the signal handling to interrupt, will trigger failure.

File: worker.go [hunk 9]
```diff
@@ -201,7 +209,31 @@ func drainWorkerThreads() []*phpThread {
...
+	case <-done:
+		// everyone yielded in time
+	case <-time.After(drainGracePeriod):
+		for _, thread := range drainedThreads {
+			if !thread.state.Is(state.Yielding) {
+				C.frankenphp_force_kill_thread(thread.forceKill)
+			}
+		}
+		if globalLogger.Enabled(globalCtx, slog.LevelWarn) {
+			globalLogger.LogAttrs(globalCtx, slog.LevelWarn, "worker threads did not yield within grace period, force-killing stuck threads")
```
Comment (withinboredom) [inline worker.go]: This gets logged AFTER we've sent the force-kill, not before. Thus if the process crashes during kill ... nobody will know why. I'd suggest moving this log line before the actual effect.

File: frankenphp.c [hunk 10]
```diff
@@ -92,6 +92,115 @@ static bool is_forked_child = false;
...
+static pthread_once_t kill_signal_handler_installed = PTHREAD_ONCE_INIT;
+static void install_kill_signal_handler(void) {
+  /* Install the no-op handler process-wide with SA_RESTART cleared so
+   * blocking syscalls return EINTR when the signal is delivered rather
+   * than being transparently restarted by libc. */
+  struct sigaction sa;
+  memset(&sa, 0, sizeof(sa));
+  sa.sa_handler = frankenphp_kill_signal_handler;
+  sigemptyset(&sa.sa_mask);
+  sa.sa_flags = 0;
```
Comment (withinboredom) [inline frankenphp.c]: `sa.sa_flags = SA_ONSTACK`? Go docs specify this must be set IIRC. Even though this is basically pthread to pthread, its basically defense from getting a signal from the Go side about it.

File: frankenphp.c
Comment (withinboredom) [inline frankenphp.c:1283]: What happens if we deliver the signal right after this? Is that possible?

File: frankenphp.c [hunk 11]
```diff
@@ -92,6 +92,115 @@ static bool is_forked_child = false;
...
+#endif
+
+#ifdef FRANKENPHP_HAS_KILL_SIGNAL
+/* No-op handler: signal delivery is sufficient on its own because it
+ * forces the in-flight syscall to return EINTR. The VM then observes
+ * vm_interrupt on the next opcode and unwinds via zend_timeout(). */
+static void frankenphp_kill_signal_handler(int sig) { (void)sig; }
+
+static pthread_once_t kill_signal_handler_installed = PTHREAD_ONCE_INIT;
+static void install_kill_signal_handler(void) {
```
Comment (withinboredom) [inline frankenphp.c:127]: Might be worth documenting in a comment (and in documentation that this signal number is reserved) that if this is embedded in any other go project (other than Caddy), FRANKENPHP_KILL_SIGNAL should not be used or needs to be patched.

Also, worth noting that if a php script registers a signal handler via pcntl, it will clobber this, and receive the kill signal instead of this handler. Might be worth checking github for this signal number and checking to make sure no popular libraries in PHP are already using this.

**Review by withinboredom — COMMENTED** [review]
Looks pretty good. FYI: bugs like https://github.com/php/php-src/issues/21267 mean that sometimes JIT just will never hit these vm breakpoints. So, be prepared for bug reports that aren't related to this change, but are due to JIT.

**Comment by dunglas** [comment]
Good work! Shouldn't this code be directly on php-src (TSRM)? It could be useful in other contexts than FrankenPHP.

**Comment by nicolas-grekas** [comment]
@dunglas dunno in the future, but at the moment, this allows providing the capablity to all versions of PHP, without having to wait for an hypothetical merge in 8.6+.

File: phpthread.go [hunk 12]
```diff
@@ -203,10 +261,69 @@ func go_frankenphp_after_script_execution(threadIndex C.uintptr_t, exitStatus C.
...
+// released its state (the abandon-on-shutdown path: Shutdown timed out
+// waiting for the thread, phpThreads was nilled, but the blocked syscall
+// eventually unwound). Late callbacks simply drop their work instead of
+// panicking on a nil-slice index.
+func threadForLateCallback(threadIndex C.uintptr_t) *phpThread {
+	if phpThreads == nil || int(threadIndex) >= len(phpThreads) {
+		return nil
+	}
+	return phpThreads[threadIndex]
+}
```
Comment (AlliBalliBaba) [inline phpthread.go]: This is probably not a good idea, phpThreads[threadIndex] should always return something, otherwise you'll have to handle it being nil everywhere. 

If we truly want to 'zombify' threads safely, one solution might be to set the thread_index of the actual thread to `0` and just put a 'Dummy thread' on the 0 position always

Comment (nicolas-grekas) [reply phpthread.go]: yes, I'm still working on this!

**Comment by henderkes** [comment]
> @dunglas dunno in the future, but at the moment, this allows providing the capablity to all versions of PHP, without having to wait for an hypothetical merge in 8.6+.

The two don't exclude each other, it would actually help getting this upstreamed because it will open the doors to more calls being switched to alertable. Once that lands in master, we can backport it in FrankenPHP for 8.4+.

File: frankenphp.c (hunk 11 above)
Comment (nicolas-grekas) [reply frankenphp.c:127]: Looks like this is rare: https://github.com/search?q=%22SIGRTMIN%2B3%22+language%3APHP+&type=code
Comment added in the code.

**Comment by nicolas-grekas** [comment]
PR should be ready! It took me a while to get it correct. PR description is updated. Lots of comments in the patch; let me know if that's too much.

**Comment by nicolas-grekas** [comment]
PR ready twice 😅 
I removed the "abandon" trial, that's a dead end. This is now really just a best effort and we wait for threads to stop on their own after signaling them (like before).

File: frankenphp.c [hunk 13]
```diff
@@ -1158,12 +1279,20 @@ static void *php_thread(void *arg) {
...
+   * would hand a fresh pthread a phpThreads slice already untracked.
+   * Notify Go either way so phpThread.shutdown's WaitFor(state.Done)
+   * unblocks - force-kill bails the thread through this path. */
+  if (zend_atomic_bool_load(&shutdown_in_progress)) {
+    frankenphp_log_message(
+        "Unhealthy thread unwinding after Shutdown; not restarting",
+        LOG_WARNING);
+    go_frankenphp_on_thread_shutdown(thread_index);
+    return NULL;
+  }
```
Comment (AlliBalliBaba) [inline frankenphp.c]: Do we really need the 'shutdown in progress' logic? Reaching here is already an edge case if something goes very wrong in an extension.

And just restarting and stopping the the thread again is also probably safe to do. So I'm not sure which scenario we're covering here or how to even test that this fixes the scenario.

Might be better to just remove the `shutdown_in_progress` logic entirely.

File: frankenphp.c [hunk 14]
```diff
@@ -92,6 +92,107 @@ static bool is_forked_child = false;
...
+#elif defined(PHP_WIN32)
+  if (!DuplicateHandle(GetCurrentProcess(), GetCurrentThread(),
+                       GetCurrentProcess(), &slot.thread_handle, 0, FALSE,
+                       DUPLICATE_SAME_ACCESS)) {
+    /* On failure, force_kill falls back to atomic-bool only. */
+    slot.thread_handle = NULL;
+  }
+#endif
+  go_frankenphp_store_force_kill_slot(idx, slot);
+}
```
Comment (AlliBalliBaba) [inline frankenphp.c]: I think this function is currently unused

File: frankenphp.c (hunk 13 above)
Comment (nicolas-grekas) [reply frankenphp.c]: agreed, removed!

File: frankenphp.c (hunk 14 above)
Comment (nicolas-grekas) [reply frankenphp.c]: logic inlined

**Comment by AlliBalliBaba** [comment]
Looks largely good to me, have you also tried that this unblocks actual syscalls like `system('sleep 45');` If this works, we'll probably also use the mechanism for a custom request timeout. 

It's probably good to have this on restarts, but on general shutdown people might want to have a longer grace period. Eg you might want to have some worker cleanup happening on shutdown.

**Comment by nicolas-grekas** [comment]
COol :)

> system('sleep 45');

I didn't try but this should work; it's all syscall-based.

For the longer grace period on shutdown vs restart:

The worker side can't tell the two modes apart though, can it? `frankenphp_handle_request()` returns false on both restart and shutdown, so any worker cleanup logic runs identically in both cases. I bumped the grace period from 5s to 30s so cleanup has more headroom in both paths. Let me know if that's good enough for you.

🚀

**Review by alexandre-daubois — APPROVED** [review]
Looks good!

File: export_test.go [hunk 15]
```diff
@@ -0,0 +1,12 @@
+package frankenphp
+
+import "time"
+
+// SetDrainGracePeriodForTest overrides drainGracePeriod for tests that need
+// to exercise the force-kill path without paying the production wait.
+// Returns the previous value so tests can restore it.
+func SetDrainGracePeriodForTest(d time.Duration) time.Duration {
```
Comment (dunglas) [inline export_test.go]: A lighter/cleaner option would be to just move `TestRestartWorkersForceKillsStuckThread` in a dedicated test file in the `frankenphp` package, changing this variable directly.

File: frankenphp.c
Comment (dunglas) [inline frankenphp.c:1161]: Nit: actually, we should revert this and add blank lines before other returns in a WSL-like style: https://github.com/bombsimon/wsl

File: worker.go [hunk 16]
```diff
@@ -165,16 +165,18 @@ func newWorker(o workerOpt) (*worker, error) {
...
 
-// EXPERIMENTAL: DrainWorkers finishes all worker scripts before a graceful shutdown
+// drainGracePeriod: time to wait for threads to yield before arming force-kill.
+var drainGracePeriod = 30 * time.Second
+
+// EXPERIMENTAL: DrainWorkers initiates a graceful drain of all worker scripts.
+// Blocks until every drained thread yields. Force-kill is armed after a
+// grace period to wake threads parked in blocking syscalls (sleep, I/O).
 func DrainWorkers() {
-	_ = drainWorkerThreads()
```
Comment (dunglas) [inline worker.go:170]: To keep, to prevent some lint errors

**Comment by nicolas-grekas** [comment]
Thanks Copilot, replies inline:

- **`worker_test.go`** (path style) — skipping. Every other test in the file uses `cwd + "/testdata/"`; following project convention. Go's filepath layer accepts forward slashes on Windows.
- **`worker.go`** (30s vs 5s) — good catch, PR description updated.
- **`frankenphp.h`** (macOS `SIGRTMIN`) — macOS doesn't define `SIGRTMIN` (POSIX RT signals are Linux/BSD only), so `defined(SIGRTMIN)` already excludes it. The atomic-bool-only path is what kicks in there.
- **`frankenphp.c`** (`sigaction()` return) — fixed: added a `kill_signal_handler_active` flag that's only set when `sigaction()` returns 0; `force_kill_thread` now skips `pthread_kill` if the handler isn't active so an install failure can't deliver an unhandled `SIGRTMIN+3`.

**Comment by nicolas-grekas** [comment]
@dunglas review:

- **`export_test.go`** — moved `TestRestartWorkersForceKillsStuckThread` into a new internal-package file [`worker_internal_test.go`](https://github.com/nicolas-grekas/frankenphp/blob/force-kill-primitives/worker_internal_test.go); deleted `export_test.go`. Test pokes `drainGracePeriod` directly now.
- **`worker.go`** — restored `_ = drainWorkerThreads()` in `DrainWorkers` for the lint signal.
- **`frankenphp.c`** — reverted the inlining: `frankenphp_register_thread_for_kill` is back as a static helper, with a blank line before the early `return` in `frankenphp_force_kill_thread` (WSL-style).

**Comment by dunglas** [comment]
Thank you @nicolas-grekas, impressive work
